org-code-javabuilder/lib/src/main/java/org/code/javabuilder/UserCodeCompiler.java

```diff
@@ -89,4 +92,107 @@ private CompilationTask getCompilationTask(DiagnosticCollector<JavaFileObject> d
     // create compilation task
     return compiler.getTask(null, fileManager, diagnostics, optionList, null, files);
   }
+
+  /**
+   * For specific compiler errors we want to return custom messages. Check if the given diagnostic
+   * is due to one of those errors and return the more specific message, or null if not.
+   *
+   * @param diagnostic
+   * @return custom compiler error or null
+   */
+  private String getCustomCompilerError(Diagnostic<? extends JavaFileObject> diagnostic) {
+    // Check if the compiler error is due to the student importing java.lang.System
+    // directly and give a more helpful message.
+    if (diagnostic.getCode().equals("compiler.err.already.defined.single.import")
+        && diagnostic.getMessage(Locale.US).contains("org.code.lang.System")) {
+      return "Import of java.lang.System is not supported.";
+    }
+    return null;
+  }
+
+  /**
+   * Get the compiler error as a String from the given diagnostic. We create a compiler error here
+   * because we need to subtract 1 from the line number due to our auto-import of System.
+   *
+   * <p>The compiler error format is:
+   *
+   * <pre>
+   * FileName:line number: error/warning: {error type}
+   * {code snippet}
+   * ^ (caret pointing to location of error)
+   * {error details if applicable}
+   * </pre>
+   *
+   * <p>Example:
+   *
+   * <pre>
+   * /SystemTest.java:26: error: cannot find symbol
+   *     System.out.println(System.currenTimeMillis());
+   *                              ^
+   *   symbol:   method currenTimeMillis()
+   *   location: class org.code.lang.System
+   * </pre>
+   *
+   * @param diagnostic
+   * @return compiler error String
+   */
+  private String getCompilerError(Diagnostic<? extends JavaFileObject> diagnostic) {
```

Comment:
Might be nice to add a test case for this to make sure the format outputs correctly? But also looks like we don't have any unit tests for `UserCodeCompiler`, so that can also be a future tech debt item

Reply:
I'll add a tech debt item, I have a feeling it's a fairly big undertaking because we have to mock quite a few things.

Reply:
https://codedotorg.atlassian.net/browse/JAVA-467